### PR TITLE
61 - Implement new Resource API with LruCache and AJAX methods

### LINKED ISSUE #61: Resource - LRU Cache
Make resource LRU Cache (dbly LL + map)

Configuration by size.

### PR DESCRIPTION
**Pull request by chikeichan**
This resolves #61 

2 main features:
 - Resource.cache is now a LruCache. 
 - new AJAX class surfacing 4 new intercept functons to intercept network requests. These should be brought down to two in the future.

### CODE REVIEW

File: src/helpers/ajax.js [hunk 1]
```diff
@@ -1,31 +1,51 @@
...
+    return res;
+};
 
-    if (opts.encode) {
-        opts.url += encodeURI(opts.encode(opts.payload));
-    }
+var Ajax = function(opts) {
+    var xhr        = new XMLHttpRequest();
+    xhr.onload  = function() {
+        var response = GLOBAL_AJAX_RESPONSE_INTERCEPT(xhr);
```
Comment (chikeichan) [inline src/helpers/ajax.js]: This will intercept all network response

File: src/helpers/ajax.js [hunk 2]
```diff
@@ -1,31 +1,51 @@
...
+        opts.onload(response);
+    };
 
-    xhr.open(opts.type.toUpperCase(), opts.url);
-    xhr.setRequestHeader('Content-Type', opts.contentType);
+    this.url     = opts.url;
+    this.type    = opts.type.toUpperCase();
+    this.headers = {};
+    this.xhr     = xhr;
+    GLOBAL_AJAX_REQUEST_INTERCEPT(this);
```
Comment (chikeichan) [inline src/helpers/ajax.js]: This will intercept all network request

File: src/resource/resource.js [hunk 3]
```diff
@@ -1,63 +1,167 @@
...
+    });
+
+    this.interceptRequest(request);
+    request.send(opts.payload);
+};
+
+Resource._constructor.prototype.interceptResponse = function(res) { return res; };
+Resource._constructor.prototype.interceptRequest = function(req) { return req; };
+
+Resource._constructor.prototype.encodeUrlParam = function(param) {
```
Comment (chikeichan) [inline src/resource/resource.js]: Helper method surfaced to help with common url param encoding

File: src/resource/resource.js [hunk 4]
```diff
@@ -1,63 +1,167 @@
...
+Resource.extend = extend;
+
+//////////////////////////////////////////////
+////////////////// LRU Cache /////////////////
+//////////////////////////////////////////////
+
+var LruCache = function(size) {
+    var map = {};
+    var list = new DblyLinkedList();
+    var maxLength = size || 0;
```
Comment (chikeichan) [inline src/resource/resource.js]: this.set(id, value) will store value to cache at key: id, also pushing node to front.
this.get(id) will get value from cache at key: id, also pushign node to front.

Whenever cache size exceed maxLength, the least recently used node will be dropped. When maxLength is set to 0 (default value), nothing will be dropped.

**Comment by sdtsui** [comment]
Looks good. Both of these PRs are making more sense now. Please resubmit after a rebase.

If you have an article to link to, regarding:
What `GLOBAL_AJAX_RESPONSE_INTERCEPT` is
and how it works in your `ajax` function, please send.

Otherwise I will google and ping for an async discussion next week.

**Comment by chikeichan** [comment]
Rebased. Basically i will to expose the whole lifecycle of the ajax request for dev.

So each request will go thru -
Global Request Interceptor
Local Request Interceptor
Send Request
Receive Response
Global Response Interceptor
Local Response Interceptor

**Pull request merged**